sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -297,7 +297,243 @@ abstract class RapidsShuffleThreadedWriterBase[K, V](
+          } else {
+            // Sleep briefly to avoid busy waiting
+            Thread.sleep(1)
+          }
+        }
+      }
+    }
+    
+    val writerFuture = writerExecutor.submit(writerTask)
+
+    try {
+      
+      while (records.hasNext) {
+        val record = records.next()
+        val key = record._1
+        val value = record._2
+        val reducePartitionId: Int = partitioner.getPartition(key)
+        recordsWritten += 1
+
+        // Get or create futures queue for this partition
+        val futures = partitionFutures.computeIfAbsent(reducePartitionId, 
+          _ => new java.util.concurrent.ConcurrentLinkedQueue[Future[Long]]())
+
+
+        val (cb, recordSize) = value match {
+          case columnarBatch: ColumnarBatch =>
+            if (columnarBatch.numCols() > 0) {
+              columnarBatch.column(0) match {
+                case _: SlicedGpuColumnVector =>
+                  (SlicedGpuColumnVector.incRefCount(columnarBatch),
+                    SlicedGpuColumnVector.getTotalHostMemoryUsed(columnarBatch))
+                case _: SlicedSerializedColumnVector =>
+                  (SlicedSerializedColumnVector.incRefCount(columnarBatch),
+                    SlicedSerializedColumnVector.getTotalHostMemoryUsed(columnarBatch))
+                case _ =>
+                  (null, 0L)
+              }
+            } else {
+              (columnarBatch, 0L)
+            }
+          case _ =>
+            (null, 0L)
+        }
+
+        // Acquire limiter and process compression task immediately
+        val waitOnLimiterStart = System.nanoTime()
+        limiter.acquireOrBlock(recordSize)
+        waitTimeOnLimiterNs += System.nanoTime() - waitOnLimiterStart
+        
+        // Submit compression task using queueWriteTask to
+        // ensure same partition tasks run serially
+        val slotNum = RapidsShuffleInternalManagerBase.getNextWriterSlot
+        val future = RapidsShuffleInternalManagerBase.queueWriteTask(slotNum, () => {
+          withResource(cb) { _ =>
+            try {
+              // Get or create buffer for this partition
+              val buffer = partitionBuffers.computeIfAbsent(reducePartitionId,
+                _ => new ByteArrayOutputStream())
+
+              // Serialize + compress to memory buffer
+              val compressedOutputStream = blockManager.serializerManager.wrapStream(
+                ShuffleBlockId(shuffleId, mapId, reducePartitionId), buffer)
+
+              val serializationStream = serializerInstance.serializeStream(compressedOutputStream)
+              withResource(serializationStream) { serializer =>
+                serializer.writeKey(key.asInstanceOf[Any])
+                serializer.writeValue(value.asInstanceOf[Any])
+              }
+
+              // Track total data size
+              totalDataSize.addAndGet(recordSize)
+              recordSize // Return record size
+            } finally {
+              limiter.release(recordSize)
```

Comment:
We are keeping the host buffer after the task completes, we can't release the limiter here.

We instead would want to release the limiter when we close `buffer`.